dtm0/drlink.c

```diff
@@ -0,0 +1,532 @@
+	M0_PRE(req != NULL);
+	M0_PRE(m0_fid_is_valid(tgt));
+
+	reqh = svc->dos_generic.rs_reqh;
+	mach = m0_reqh_rpc_mach_tlist_head(&reqh->rh_rpc_machines);
+
+	owned_req = dtm0_req_fop_dup(req);
+	if (owned_req == NULL)
+		return M0_ERR(-ENOMEM);
+
+	fop = m0_fop_alloc(&dtm0_req_fop_fopt, owned_req, mach);
+	if (fop == NULL) {
+		dtm0_req_fop_fini(owned_req);
+		m0_free(owned_req);
+		return M0_ERR(-ENOMEM);
+	}
+
+	/*
+	 * When ACK is not required, the FOM may be released before
+	 * the received callback is triggered.
+	 * See ^1 in ::dtm0_rlink_rpc_item_reply_cb.
+	 */
+	fop->f_opaque = wait_for_ack ? fom : NULL;
+
+	m0_fom_init(&fom->df_gen, &drlink_fom_type, &drlink_fom_ops,
+		    NULL, NULL, reqh);
+
+	/* TODO: can we use fom->fo_fop instead? */
+	fom->df_rfop =  fop;
+	fom->df_svc  =  svc;
+	fom->df_tgt  = *tgt;
+	fom->df_wait_for_ack = wait_for_ack;
+	fom->df_parent_sm_id = m0_sm_id_get(&parent_fom->fo_sm_phase);
+
+	m0_co_context_init(&fom->df_co);
+	m0_co_op_init(&fom->df_co_op);
+
+	M0_LEAVE();
+
+	return M0_RC(0);
+}
+
+static void drlink_fom_fini(struct m0_fom *fom)
+{
+	struct drlink_fom *df = fom2drlink_fom(fom);
+	m0_fop_put_lock(df->df_rfop);
+	m0_co_op_fini(&df->df_co_op);
+	m0_fom_fini(fom);
+	m0_free(fom);
+}
+
+static void co_long_write_lock(struct m0_co_context     *context,
+			       struct m0_long_lock      *lk,
+			       struct m0_long_lock_link *link,
+			       int                       next_phase)
+{
+	int outcome;
+	M0_CO_REENTER(context);
+	outcome = M0_FOM_LONG_LOCK_RETURN(m0_long_write_lock(lk, link,
+							     next_phase));
+	M0_CO_YIELD_RC(context, outcome);
+}
+
+static void co_rpc_link_connect(struct m0_co_context *context,
+				struct m0_rpc_link   *rlink,
+				struct m0_fom        *fom,
+				int                   next_phase)
+{
+	M0_CO_REENTER(context);
+
+	m0_chan_lock(&rlink->rlk_wait);
+	m0_fom_wait_on(fom, &rlink->rlk_wait, &fom->fo_cb);
+	m0_chan_unlock(&rlink->rlk_wait);
+
+	m0_rpc_link_connect_async(rlink, M0_TIME_NEVER, NULL);
+	m0_fom_phase_set(fom, next_phase);
+
+	M0_CO_YIELD_RC(context, M0_FSO_WAIT);
+}
+
+static int find_or_add(struct m0_dtm0_service *dtms,
+		       const struct m0_fid    *tgt,
+		       struct dtm0_process   **out)
+{
+	struct dtm0_process *process;
+	int                  rc = 0;
+
+	M0_ENTRY();
+	M0_PRE(m0_mutex_is_locked(&dtms->dos_generic.rs_mutex));
+
+	process = dtm0_service_process__lookup(&dtms->dos_generic, tgt);
+	if (process == NULL) {
+		rc = M0_ALLOC_PTR(process) == NULL ?
+			M0_ERR(-ENOMEM) :
+			dtm0_process_init(process, dtms, tgt);
+		if (rc != 0)
+			m0_free(process);
+	}
+
+	if (rc == 0) {
+		M0_POST(process != NULL);
+		*out = process;
+	}
+	return M0_RC(rc);
+}
+
+enum drlink_fom_state {
+	DRF_INIT = M0_FOM_PHASE_INIT,
+	DRF_DONE = M0_FOM_PHASE_FINISH,
+	DRF_LOCKING = M0_FOM_PHASE_NR,
+	DRF_CONNECTING,
+	DRF_SENDING,
+	DRF_WAITING_FOR_REPLY,
+	DRF_FAILED,
+	DRF_NR,
+};
+
+static struct m0_sm_state_descr drlink_fom_states[] = {
+	/* terminal states */
```

Comment:
Move this comment to "DRF_DONE"

Reply:
makes sense, fixed in https://github.com/Seagate/cortx-motr/pull/802/commits/55385c7ee5f01d4d021d647f61f61f7a497ae224